### PR TITLE
Unmanaged pods should fail fast when evicted/preempted/deleted

### PR DESCRIPTION
**Pull request by KnVerey**
## Purpose

- Bugfix: Unmanaged pods should fail fast if they're evicted or preempted. Previously, we made pods' failure condition ignore these two statuses, because it was very occasionally making a parent's deploy fail incorrectly because the parent was unlucky enough to have all its children hit one of those conditions (which are recoverable, because the parent spawns replacement children). For unmanaged pods, these states are not recoverable and should not be ignored.
- Improvement: We should not wait around forever if a pod is deleted out of band. If we've deployed it and it doesn't exist, it isn't magically going to appear in the future. This is especially important for the runner. Note that I'm continuing to treat non-404 errors as transient. Related to https://github.com/Shopify/kubernetes-deploy/issues/347
- Improvement: same as above, but handles the window where the deletion request has been made successfully but the pod isn't actually gone yet (i.e. deletion timestamp is set). **Reviewers, do you agree with this?** The containers should at least have received SIGTERM when we see this, but they may still be running.

## Things I don't like

- I think we should try splitting `Pod` into `Pod` and `ManagedPod`. _Many_ of its methods do different things based on whether or not it has a parent. I don't think that is necessary for this PR though.
- The `disappeared?` thing should be implemented for all resources, not just this one (https://github.com/Shopify/kubernetes-deploy/issues/347). I do think this is a good start though, and it is a lot more important for pods (because of kubernetes-run).
- Because we're using kubectl to get these resources (which is _really_ difficult to change at this point), I can't check the actual response status. The "404" check is actually looking for a non-zero exit and an error string.
- I had to change a ton of the unit tests to accommodate the new `raise_on_404` option. We could probably improve the mocking strategy to make this less painful, but I just brute-forced it for now.

cc @Shopify/cloudx

### CODE REVIEW

File: test/integration/runner_task_test.rb [hunk 1]
```diff
@@ -18,7 +18,7 @@ def test_run_without_verify_result_succeeds_as_soon_as_pod_is_successfully_creat
       "Result: SUCCESS",
       "Result verification is disabled for this task",
       "The following status was observed immediately after pod creation:",
-      %r{Pod/task-runner-\w+\s+Pending},
+      %r{Pod/task-runner-\w+\s+(Pending|Running)},
```
Comment (KnVerey) [inline test/integration/runner_task_test.rb]: This change and the one below are to decrease test flakiness. It's not important to these two test cases whether the pod manages to start running in the run window.

File: lib/kubernetes-deploy/kubernetes_resource.rb [hunk 2]
```diff
@@ -121,12 +122,23 @@ def file_path
     end
 
     def sync(mediator)
-      @instance_data = mediator.get_instance(kubectl_resource_type, name)
+      @instance_data = mediator.get_instance(kubectl_resource_type, name, raise_on_404: true)
+    rescue KubernetesDeploy::Kubectl::ResourceNotFoundError
+      @disappeared = true if deploy_started?
```
[the remark this answers is not in the document]
Comment (KnVerey) [reply lib/kubernetes-deploy/kubernetes_resource.rb]: > deploy_started? is here because if the pod hasn't been created yet then the 404 is actually expected, right?

Exactly.

> I think we need to prioritize the state tracking refactor. We're effectively adding a new state here.

Are we? This is of course new data about the state of the resource in a general sense, but it isn't a new end state for the resource, which is what that refactor was about (the new state that triggered it was "ignored"). In other words, our end states are still succeeded, failed and timed out, and this is just a new way that resources can fail.

File: lib/kubernetes-deploy/kubernetes_resource.rb [hunk 3]
```diff
@@ -121,12 +122,23 @@ def file_path
...
+      @instance_data = {}
     end
 
     def after_sync
     end
 
+    def deleted?
+      @instance_data.dig('metadata', 'deletionTimestamp').present?
```
[the remark this answers is not in the document]
Comment (KnVerey) [reply lib/kubernetes-deploy/kubernetes_resource.rb]: > So this basically means that k8s has asked the pod to "please go away", but in theory the pod might still exist and the process might still be running, right? 

Yes. I should probably call this `terminating?` -- that's the terminology kubectl uses.

> What's the reasoning for not checking whether the pod has actually already been deleted? Because this will catch deletion requests earlier?

Yes. We've also seen a k8s bug in the past where resources would get stuck in the terminating state indefinitely, even after the underlying container was gone (though it hasn't been reported in recent version afaik). (Note that we are also checking that they have actually been deleted--if so `disappeared?` will be true)

File: lib/kubernetes-deploy/sync_mediator.rb [hunk 4]
```diff
@@ -10,12 +10,16 @@ def initialize(namespace:, context:, logger:)
...
-      if @cache.key?(kind)
-        @cache.dig(kind, resource_name) || {}
-      else
-        request_instance(kind, resource_name)
+    def get_instance(kind, resource_name, raise_on_404: false)
+      unless @cache.key?(kind)
+        return request_instance(kind, resource_name, raise_on_404: raise_on_404)
       end
+
+      cached_instance = @cache[kind].fetch(resource_name, {})
```
[the remark this answers is not in the document]
Comment (KnVerey) [reply lib/kubernetes-deploy/sync_mediator.rb]: The only place that can set it is this:

```ruby
      @cache[kind] = JSON.parse(raw_json)["items"].each_with_object({}) do |r, instances|
        instances[r.dig("metadata", "name")] = r
      end
```

That's way too densely written. If I rewrite it like this it is clearer that it shouldn't be possible to have a `nil` value:

```ruby
      instances = {}
      JSON.parse(raw_json)["items"].each do |resource|
        resource_name = resource.dig("metadata", "name")
        instances[resource_name] = resource
      end
      @cache[kind] = instances
```

File: test/integration/runner_task_test.rb [hunk 5]
```diff
@@ -88,6 +88,40 @@ def test_run_with_verify_result_success
...
+        if task_runner.pod_name.present?
+          begin
+            kubeclient.delete_pod(task_runner.pod_name, @namespace)
+            break
+          rescue Kubeclient::ResourceNotFoundError
+            sleep 0.1
+            retry
+          end
+        end
+        sleep 0.1
```
[the remark this answers is not in the document]
Comment (KnVerey) [reply test/integration/runner_task_test.rb]: It's a tiny throttle that takes effect between when we start the thread and when the pod name is generated. The on one L102 takes effect between when the pod name is generated and when the the pod has been created.

File: test/integration/runner_task_test.rb [hunk 6]
```diff
@@ -88,6 +88,40 @@ def test_run_with_verify_result_success
...
+    result = task_runner.run(run_params(log_lines: 20, log_interval: 1))
+    assert_task_run_failure(result)
+
+    assert_logs_match_all([
+      "Pod creation succeeded",
+      "Result: FAILURE",
+      /Pod status\: (Terminating|Disappeared)/,
+    ])
+  ensure
+    if deleter_thread
```
[the remark this answers is not in the document]
Comment (KnVerey) [reply test/integration/runner_task_test.rb]: Good question. This is for the case where the test is failing ultimately because of a problem in the thread (which I originally had when I wrote this). If the thread raises and you never join it, you'll never see the error and the test will suck to debug.

**Comment by KnVerey** [comment]
I've pushed a commit to address the comments. Please take a look.

> My big concern here is using exceptions as control flow. But, I don't see a better way.

Are you talking about `ResourceNotFoundError`? It's not really changing the control flow so much as passing a piece of data; the consequence is setting `@disappeared`, not changing what we do next. Another way to do that would be to make both `Kubectl#run` and `SyncMediator.get_instance` able to pass back enhanced information (e.g. multiple values, or some new type of object). That would be a huge change, and I don't really see why it would be better. It would also be possible to have it return `nil` instead of `{}` and have that mean "not found" (other errors would still have to return {}), but that feels a bit sketchy to me and handling those possible nils everywhere would not be fun. 

tl;dr I haven't thought of an alternative I like better either. Can you be more specific about your concern? Having a resource fetcher raise an exception when the target is missing doesn't seem all that weird to me tbh.

File: lib/kubernetes-deploy/kubernetes_resource.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (dturn) [reply lib/kubernetes-deploy/kubernetes_resource.rb]: I still disagree. The refactor is about mutually exclusive states not just terminal ones.

File: test/integration/runner_task_test.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (dturn) [reply test/integration/runner_task_test.rb]: Should this be in an else block?

File: test/integration/runner_task_test.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (dturn) [reply test/integration/runner_task_test.rb]: Can you leave a comment like: `join to ensure error message is present is printed`? Also do we want to set a limit on how long we'll wait for the join?

File: lib/kubernetes-deploy/kubernetes_resource/pod.rb [hunk 7]
```diff
@@ -86,7 +82,7 @@ def failure_message
           container_problems += "> #{red_name}: #{c.doom_reason}\n"
         end
       end
-      "#{phase_problem}#{container_problems}".presence
+      "#{phase_failure_message} #{container_problems}".lstrip.presence
```
[the remark this answers is not in the document]
Comment (KnVerey) [reply lib/kubernetes-deploy/kubernetes_resource/pod.rb]: That turned out to open a small can of worms. My last commit centralizes the chunking of the debug message (into that method itself) and adds a test displaying/proving the desired output.

File: test/integration/runner_task_test.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (KnVerey) [reply test/integration/runner_task_test.rb]: Added the comment. Thinking about this more, I think the better thing to do is to set `abort_on_exception` to the thread, and not join here.
